vllm/worker/cpu_model_runner.py

```diff
@@ -104,65 +104,289 @@ def from_broadcasted_tensor_dict(
+                modality: placeholder_map.index_map()
+                for modality, placeholder_map in
+                input_data.multi_modal_placeholder_maps.items()
+            }
+
+        attn_metadata = self.runner.attn_backend.make_metadata(
+            chunked_prefill=self.chunked_prefill,
+            seq_lens=prefill_seq_lens,
+            seq_lens_tensor=seq_lens_tensor,
+            max_query_len=max_query_len,
+            max_kv_len=max_kv_len,
+            query_start_loc=query_start_loc,
+            kv_start_loc=kv_start_loc,
+            max_decode_seq_len=input_data.max_decode_seq_len,
+            num_prefills=input_data.num_prefills,
+            num_prefill_tokens=input_data.num_prefill_tokens,
+            num_decode_tokens=input_data.num_decode_tokens,
+            block_tables=block_tables,
+            prefill_block_tables=prefill_block_tables,
+            slot_mapping=slot_mapping,
+            multi_modal_placeholder_index_maps=placeholder_index_maps,
+        )
 
         return self.model_input_cls(
             input_tokens=input_tokens,
             input_positions=input_positions,
+            seq_lens=input_data.seq_lens,
+            query_lens=input_data.query_lens,
             attn_metadata=attn_metadata,
             multi_modal_kwargs=multi_modal_kwargs,
-            # query_lens is not needed if chunked prefill is not
-            # supported. Since CPU worker doesn't support chunked prefill
-            # just use seq_lens instead.
-            seq_lens=seq_lens,
-            query_lens=seq_lens,
         )
 
-    def _compute_multi_modal_input(
-        self,
-        seq_data: SequenceData,
-        computed_len: int,
-        seq_group_metadata: SequenceGroupMetadata,
-    ):
+    def _build_input_data(self):
+        for seq_group_metadata in self.seq_group_metadata_list:
+            for seq_id, seq_data in seq_group_metadata.seq_data.items():
+                self._compute_input_tokens(self.input_data, seq_group_metadata,
+                                           seq_data, seq_id)
+                if (seq_group_metadata.is_prompt
+                        and seq_group_metadata.multi_modal_data):
+                    self._compute_multi_modal_input(seq_group_metadata,
+                                                    seq_data)
+
+    def _compute_input_tokens(self, data: ModelInputData,
+                              seq_group_metadata: SequenceGroupMetadata,
+                              seq_data: SequenceData, seq_id: int):
+        """
+        Compute input tokens, positions, block table and slot mapping.
+        """
+        is_prompt = seq_group_metadata.is_prompt
+        token_chunk_size = seq_group_metadata.token_chunk_size
+        block_size = self.runner.block_size
+
+        block_table = seq_group_metadata.block_tables[seq_id]
+        seq_len = seq_data.get_len()
+        context_len = seq_data.get_num_computed_tokens()
+        if is_prompt:
```

Comment:
Can we separate these to two functions like `_compute_prompt` and `_compute_decode`? I think this if else statement is a little bit too long.

Reply:
Updated